Pass process group info into NcclWork

shengfukevin opened the pull request:
Summary: Pass process group info into NcclWork

Test Plan: buck2 run mode/dev-nosan kineto/libkineto/fb/integration_tests:pytorch_execution_trace_integration_test

Differential Revision: D61677160


cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -725,7 +731,7 @@ void ProcessGroupNCCL::WorkNCCL::synchronizeInternal(
   RECORD_PARAM_COMMS(
       static_cast<int>(this->seq_), // seq
-      std::make_tuple("", ""), // PG name tuple
+      std::make_tuple(pgName_, pgDesc_), // PG name tuple
```
just for my understanding, is the motivation of this change so that (pgName_, pgDesc) can be logged from this line?

also, do we need to make these copies? i thought work has a reference to parent PG, but maybe not?

shengfukevin replied:
Yes, I am working on PyTorch execution trace, "wait" op missed the process group info, so it does not know which collective to wait for. 

I can not find a reference to parent PG. Let me know if you find it.

wconstab replied:
cc @fduwjj @shuqiangzhang Lmk if you guys want to change this later, it might be generally useful to make work hold a ref to parent.  But ill stamp for now.

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -261,6 +261,8 @@ class TORCH_API ProcessGroupNCCL : public Backend {
 
     // Constructor takes a list of CUDA devices
     WorkNCCL(
```
Can you make `WorkNCCL` a friend class of `ProcessGroupNCCL` so that you don't need to pass in these two. cc: @kwen2501 who proposed this idea originally.

shengfukevin replied:
WorkNCCL does not have a reference to ProcessGroupNCCL, even make ProcessGroupNCCL a friend class of WorkNCCL, WorkNCCL still can not access pgName_ and pgDesc_.

GSSBMW commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -329,6 +331,12 @@ class TORCH_API ProcessGroupNCCL : public Backend {
 
    protected:
+    // The process group name
+    std::string pgName_;
```
How about using name `pgUID` to keep consistent with `ProcessGroupNCCL` calss? Seems its previous name is `pg_name_`, but is changed to `pg_uid_` now.

shengfukevin replied:
fixed.

shengfukevin committed:
Pass process group info into NcclWork (#134269)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/134269

Pass process group info into NcclWork

Test Plan: buck2 run mode/dev-nosan kineto/libkineto/fb/integration_tests:pytorch_execution_trace_integration_test

Reviewed By: wconstab

Differential Revision: D61677160
